features/include/pcl/features/impl/gasd.hpp

```diff
@@ -0,0 +1,417 @@
+    copyShapeHistogramsToOutput (shape_grid_size, shape_hists_size_, shape_hists, output, pos_);
+
+    // set remaining values of the descriptor to zero (if any)
+    std::fill (output.points[0].histogram + pos_, output.points[0].histogram + output.points[0].descriptorSize (), 0.0f);
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDColorEstimation<PointInT, PointOutT>::copyColorHistogramsToOutput (const size_t grid_size,
+                                                                         const size_t hists_size,
+                                                                         std::vector<Eigen::VectorXf> &hists,
+                                                                         PointCloudOut &output,
+                                                                         size_t &pos)
+  {
+    for (size_t i = 0; i < grid_size; ++i)
+    {
+      for (size_t j = 0; j < grid_size; ++j)
+      {
+        for (size_t k = 0; k < grid_size; ++k)
+        {
+          const size_t idx = ( (i + 1) * (grid_size + 2) + (j + 1)) * (grid_size + 2) + (k + 1);
+
+          hists[idx][1] += hists[idx][hists_size + 1];
+          hists[idx][hists_size] += hists[idx][0];
+
+          std::copy (hists[idx].data () + 1, hists[idx].data () + hists_size + 1, output.points[0].histogram + pos);
+          pos += hists_size;
+        }
+      }
+    }
+  }
+
+  //////////////////////////////////////////////////////////////////////////////////////////////
+  template <typename PointInT, typename PointOutT> void
+  GASDColorEstimation<PointInT, PointOutT>::computeFeature (PointCloudOut &output)
+  {
+    // call shape feature computation
+    GASDEstimation<PointInT, PointOutT>::computeFeature (output);
+
+    const size_t color_grid_size = color_half_grid_size_ * 2;
+
+    // each histogram dimension has 2 additional bins, 1 in each boundary, for performing interpolation
+    std::vector < Eigen::VectorXf
+        > color_hists ( (color_grid_size + 2) * (color_grid_size + 2) * (color_grid_size + 2), Eigen::VectorXf::Zero (color_hists_size_ + 2));
+
+    // for each sample
+    for (size_t i = 0; i < shape_samples_.size (); ++i)
+    {
+      // compute shape histogram array coord based on distance between sample and centroid
+      const Eigen::Vector4f p (shape_samples_[i].x, shape_samples_[i].y, shape_samples_[i].z, 0.0f);
+
+      // compute hue value
+      float hue;
+
+	  const unsigned char max = std::max (shape_samples_[i].r, std::max (shape_samples_[i].g, shape_samples_[i].b));
+      const unsigned char min = std::min (shape_samples_[i].r, std::min (shape_samples_[i].g, shape_samples_[i].b));
+
+      if (max == 0) // division by zero
+      {
+        hue = 0.f;
+      }
+      else
+      {
+        if (min == max) // diff == 0 -> division by zero
+        {
+          hue = 0;
+        }
```

Comment:
I'm not sure if these two checks you do above are sufficient to ensure that `1/diff` will not yield a division by zero. They're arithmetically correct, it's just the floating point precision specificities which concern me. Do the following instead.

```cpp
float hue = 0.f;

const float diff_inv = 1.f / static_cast <float> (max - min);

if (pcl_isfinite (diff_inv))
{
  if (max == shape_samples_[i].r)
  {
    hue = 60.f * (static_cast <float> (shape_samples_[i].g - shape_samples_[i].b) * diff_inv);
  }
  else if (max == shape_samples_[i].g)
  {
    hue = 60.f * (2.f + static_cast <float> (shape_samples_[i].b - shape_samples_[i].r) * diff_inv);
  }
  else
  {
    hue = 60.f * (4.f + static_cast <float> (shape_samples_[i].r - shape_samples_[i].g) * diff_inv); // max == b
  }

  if (hue < 0.f)
  {
    hue += 360.f;
  }
}
```